src/base/webrtc-provider.ts

```diff
@@ -0,0 +1,62 @@
+import { EventEmitter } from "events";
+import StrictEventEmitter from "strict-event-emitter-types";
+import { IStream } from "./stream-provider";
+
+/**
+ * Possible event definitions for the webrtc provider
+ */
+interface IWebrtcProviderEvents {
+  /**
+   * Emitted when an error occurs
+   */
+  error: (err: any) => void;
+
+  /**
+   * Emitted when we connect on an rtc level
+   */
+  connect: () => void;
+
+  /**
+   * Emitted when we disconnect on an rtc level
+   */
+  disconnect: () => void;
+
+  /**
+   * Emitted when signal data is prepared for transmission
+   */
+  signal: (data: any) => void;
+
+  /**
+   * Emitted when webrtc peer-sent data has arrived
+   */
+  data: (data: any) => void;
+}
+
+/**
+ * The strongly-typed webrtc emitter type that we'll extend for our provider
+ */
+type WebrtcEmitter = StrictEventEmitter<EventEmitter, IWebrtcProviderEvents>;
+
+/**
+ * The base webrtc provider type
+ */
+export abstract class IWebrtcProvider extends (EventEmitter as new() => WebrtcEmitter) {
```

Comment:
the `I` prefix on the name here is a bit weird, as it's a class not an interface - However, it conveys that this class should not be used on it's own, so i've left it. If folks feel __strongly__ this can be changed.

Reply:
Yeah, I feel it's weird since the `I` prefix is most common for interface and it's consistent with the rest of our code base. How's about `BaseWebrtcProvider`? ;)

Reply:
😓 sure. I'll wait to iterate til the rest of your review lands.

Reply:
done